File: CivitasRuby/controlador.rb
require_relative 'civitas_juego'
require_relative 'vista_textual'
require_relative 'operaciones_juego'
require_relative 'gestiones_inmobiliarias'
require_relative 'operaciones_juego'
require_relative 'salidas_carcel'
require_relative 'respuestas'
require_relative 'operacion_inmobiliaria'

module Civitas
  
  class Controlador
    def initialize(juego, vista)
      @juego = juego
      @vista = vista
    end
    
    def juega
      @vista.setCivitasJuego(@juego)
      
      while not @juego.final_del_juego do
        @vista.actualizarVista
        @vista.pausa
        siguiente_operacion = @juego.siguiente_paso

        if siguiente_operacion != Operaciones_juego::PASAR_TURNO
          @vista.mostrarEventos
        end
        
        if not @juego.final_del_juego
          case siguiente_operacion
          when Operaciones_juego::COMPRAR
            if @vista.comprar == Respuestas::SI
              @juego.comprar
            end
            @juego.siguiente_paso_completado(siguiente_operacion)
        
          when Operaciones_juego::GESTIONAR
            @vista.gestionar
            i_gestion = @vista.getGestion
            ip = @vista.getPropiedad
            gestion = Lista_GestionesInmobiliarias[i_gestion]
            operacion = OperacionInmobiliaria.new(gestion, ip)
            
            case gestion
            when GestionesInmobiliarias::TERMINAR
              @juego.siguiente_paso_completado(siguiente_operacion)
            when GestionesInmobiliarias::VENDER
              @juego.vender(ip)
            when GestionesInmobiliarias::HIPOTECAR
              @juego.hipotecar(ip)
            when GestionesInmobiliarias::CANCELAR_HIPOTECA
              @juego.cancelar_hipoteca(ip)
            when GestionesInmobiliarias::CONSTRUIR_CASA
              @juego.construir_casa(ip)
            when GestionesInmobiliarias::CONSTRUIR_HOTEL
              @juego.construir_hotel(ip)
            end
              
          when Operaciones_juego::SALIR_CARCEL
            s = @vista.salir_carcel
            if s == SalidasCarcel::PAGANDO
              @juego.salir_carcel_pagando
            elsif s == SalidasCarcel::TIRANDO
              @juego.salir_carcel_tirando
            end
            
            @juego.siguiente_paso_completado(siguiente_operacion)
          else
          end
        end
      end
      
      # final del juego -> muestra ranking 
			ranking = @juego.ranking
      puts("El juego ha finalizado. La clasificacion ha sido:\n")
      for i in 0..ranking.length()-1
        cadena = i + ") " + ranking[i].to_s
        puts(cadena)
      end
    end
  end
end

File: CivitasJava/GUI/Controlador.java
package GUI;

import java.util.ArrayList;
import civitas.CivitasJuego;
import civitas.GestionesInmobiliarias;
import civitas.OperacionesJuego;
import civitas.OperacionInmobiliaria;
import civitas.Respuestas;
import civitas.SalidasCarcel;
import civitas.Jugador;

public class Controlador {
	private CivitasJuego juego;
	private CivitasView vista;

        Controlador(CivitasJuego _juego, CivitasView _vista){
            juego = _juego;
            vista = _vista;
        }
        
	void juega(){
        vista.setCivitasJuego(juego);

        while (!juego.finalDelJuego()){
            vista.actualizarVista();
            OperacionesJuego siguienteOperacion = juego.siguientePaso();
            vista.mostrarSiguienteOperacion(siguienteOperacion);
            
            if (siguienteOperacion != OperacionesJuego.PASAR_TURNO){
                vista.mostrarEventos();
            }

            if (!juego.finalDelJuego()){
                switch(siguienteOperacion){
                    case COMPRAR:
                        if (vista.comprar() == Respuestas.SI)
                            juego.comprar();

                        juego.siguientePasoCompletado(siguienteOperacion);
                        break;
                    
                    case GESTIONAR:
                        vista.gestionar();
                        int iGestion = vista.getGestionElegida();
                        int ip = vista.getPropiedadElegida();
                        GestionesInmobiliarias gestion = GestionesInmobiliarias.values()[iGestion];
                        OperacionInmobiliaria operacion = new OperacionInmobiliaria(gestion, ip);
                                
                        switch(gestion){
                            case TERMINAR:
                                juego.siguientePasoCompletado(siguienteOperacion);
                                break;
                                    
                            case VENDER:
                                juego.vender(ip);
                                break;
                                        
                            case HIPOTECAR:
                                juego.hipotecar(ip);
                                break;
                                        
                            case CANCELAR_HIPOTECA:
                                juego.cancelarHipoteca(ip);
                                break;
                                        
                            case CONSTRUIR_CASA:
                                juego.construirCasa(ip);
                                break;
                                        
                            case CONSTRUIR_HOTEL:
                                juego.construirHotel(ip);
                                break;
                                        
                            default: break;
                        }
                        break;
                                
                    case SALIR_CARCEL:
                        SalidasCarcel s = vista.salirCarcel();
                        if (s == SalidasCarcel.PAGANDO)
                            juego.salirCarcelPagando();
                        else if (s == SalidasCarcel.TIRANDO)
                            juego.salirCarcelTirando();
                        
                        juego.siguientePasoCompletado(siguienteOperacion);
                        break;
                      
                    default: break;
                }
            }
        }
		
            // final del juego
            ArrayList<Jugador> ranking = new ArrayList<Jugador>(juego.ranking());
            System.out.println("\nEl juego ha finalizado. La clasificacion ha sido:");
                for (int i = 0; i < ranking.size(); i++)
                    System.out.println(i + ") " + ranking.get(i));
	}
}
